lib/spack/spack/cmd/providers.py

```diff
@@ -35,11 +33,41 @@
 
 
 def setup_parser(subparser):
+    subparser.epilog = 'If called without argument returns ' \
+                       'the list of all valid virtual packages'
     subparser.add_argument(
-        'vpkg_spec', metavar='VPACKAGE_SPEC', nargs=argparse.REMAINDER,
-        help='find packages that provide this virtual package')
+        'virtual_package',
+        nargs='*',
+        help='find packages that provide this virtual package'
+    )
 
 
 def providers(parser, args):
-    for spec in spack.cmd.parse_specs(args.vpkg_spec):
-        colify(sorted(spack.repo.providers_for(spec)), indent=4)
+    valid_virtuals = sorted(spack.repo.provider_index.providers.keys())
+    valid_virtuals_str = 'Virtual packages:\n\t' if sys.stdout.isatty() else ''
+    valid_virtuals_str += ', '.join(valid_virtuals)
```

Comment:
Can we actually use `colify` here? Then I don't have to do any work when parsing the output. We already have a lot of virtual packages, and that number is only going to grow. It makes sense to use the same formatting style as `spack list`.